packages/@aws-containers/aws-ecs-builder/lib/extensions/appmesh.ts

```diff
@@ -0,0 +1,289 @@
+    // the maximum number of health tasks. See:
+    // https://docs.aws.amazon.com/app-mesh/latest/userguide/best-practices.html#reduce-deployment-velocity
+    if (!props.desiredCount || props.desiredCount < 4) {
+      maxPercent = 150;
+    } else {
+      maxPercent = 125;
+    }
+
+    return {
+      ...props,
+
+      // We must make sure that service tasks are registered into
+      // CloudMap so that the App Mesh proxy can find them.
+      cloudMapOptions: {
+        dnsRecordType: 'A',
+        dnsTtl: cdk.Duration.seconds(10),
+        failureThreshold: 2,
+        name: this.parentService.id,
+      },
+
+      // These specific deployment settings are currently required in order to
+      // maintain availability during a rolling deploy of the service with App Mesh
+      // https://docs.aws.amazon.com/app-mesh/latest/userguide/best-practices.html#reduce-deployment-velocity
+      minHealthyPercent: 100,
+      maxHealthyPercent: maxPercent,
+    } as ServiceBuild;
+  }
+
+  // Now that the service is defined we can create the AppMesh virtual service
+  // and virtual node for the real service
+  public useService(service: ecs.Ec2Service | ecs.FargateService) {
+    const containerextension = this.parentService.serviceDescription.get('service-container') as Container;
+
+    if (!containerextension) {
+      throw new Error('Firelens extension requires an application extension');
+    }
+
+    const cloudmapNamespace = this.parentService.cluster.defaultCloudMapNamespace;
+
+    if (!cloudmapNamespace) {
+      throw new Error('You must add a CloudMap namespace to the ECS cluster in order to use the AppMesh extension');
+    }
+
+    // Create a virtual node for the name service
+    this.virtualNode = new appmesh.VirtualNode(this.scope, `${this.parentService.id}-virtual-node`, {
+      mesh: this.mesh,
+      virtualNodeName: this.parentService.id,
+      cloudMapService: service.cloudMapService,
+      listener: {
+        portMapping: {
+          port: containerextension.trafficPort,
+          protocol: this.protocol,
+        },
+      },
+    });
+
+    // Create a virtual router for this service. This allows for retries
+    // and other similar behaviors.
+    this.virtualRouter = new appmesh.VirtualRouter(this.scope, `${this.parentService.id}-virtual-router`, {
+      mesh: this.mesh,
+      listener: {
+        portMapping: {
+          port: containerextension.trafficPort,
+          protocol: this.protocol,
+        },
+      },
+      virtualRouterName: `${this.parentService.id}`,
+    });
+
+    // Now add the virtual node as a route in the virtual router
+    this.route = this.virtualRouter.addRoute(`${this.parentService.id}-route`, {
+      routeTargets: [{
+        virtualNode: this.virtualNode,
+        weight: 1,
+      }],
+      // Ensure that the route type matches the protocol type.
+      routeType: this.protocol == appmesh.Protocol.HTTP ? appmesh.RouteType.HTTP : appmesh.RouteType.TCP,
```

Comment:
I'm confused -- do the non-http protocols mentioned [here](https://github.com/aws/aws-cdk/pull/10129/files#diff-e555d0620ac2db9b5c1924d11a90a3b0R21-R22) all map to `appmesh.RouteType.TCP`?

Reply:
For now, unfortunately. The App Mesh team is planning to add some more support for App Mesh in CDK, so this piece of code will get more full featured once App Mesh L2 constructs have support for more protocols.